WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,605 @@
+     *
+     * ERROR: Method cannot be declared public because its parameter uses an internal type
+     */
+    public func set_network_manager(_ nm: NetworkManager) {
+        network_manager = nm
+        configure()
+    }
+
+    /**
+     * Stores the input buffer of generated events in persistent storage and clears it
+     */
+    private func persist_buffer() {
+        storage_manager?.set_persisted("epc_input_buffer", input_buffer)
+    }
+
+    /**
+     * Retrieves persisted input buffer and deletes it from storage
+     *
+     * Merges retrieved events with any existing events in `input_buffer`.
+     */
+    private func recall_buffer() {
+        storage_manager?.get_persisted("epc_input_buffer", completion: {
+            value in
+            if value != nil {
+                let ib = value as! [(stream: String, schema: String, domain: String, data: [String: Any])]
+                for e in ib {
+                    input_buffer.append(e)
+                }
+                storage_manager!.del_persisted("epc_input_buffer")
+            }
+        })
+    }
+
+    /**
+     * This method is called by the application delegate in `applicationWillResignActive()` and
+     * disables event logging.
+     */
+    public func app_in_background() {
+        ENABLED = false
+    }
+    /**
+     * This method is called by the application delegate in `applicationDidBecomeActive()` and
+     * re-enables event logging.
+     *
+     * If it has been more than 15 minutes since the app entered background state, a new session is started.
+     */
+    public func app_in_foreground() {
+        ENABLED = true
+        if session_timed_out() {
+            begin_new_session()
+        }
+    }
+    /**
+     * This method is called by the application delegate in `applicationWillTerminate()`
+     *
+     * We do not persist session ID on app close because we have decided that a session ends when the
+     * user (or the OS) has closed the app or when 15 minutes of inactivity have assed.
+     */
+    public func app_will_close() {
+        ENABLED = false
+        persist_buffer()
+    }
+
+    /**
+     * Generates a new identifier using the same algorithm as EPC libraries for web and Android
+     */
+    private func generate_id() -> String {
+        var id: String = ""
+        for _ in 1...8 {
+            id += String(format: "%04x", arc4random_uniform(65535))
+        }
+        return id
+    }
+
+    /**
+     * Unset the session
+     */
+    private func begin_new_session() -> Void {
+        SESSION_ID = nil
+        cache.removeAll()
+    }
+
+    /**
+     * Generate a session identifier
+     * - Returns: session ID
+     *
+     * The identifier is a string of 20 zero-padded hexadecimal digits representing a uniformly random
+     * 80-bit integer.
+     */
+    public func session_id() -> String {
+        if SESSION_ID == nil {
+            SESSION_ID = generate_id()
+        }
+        return SESSION_ID!
+    }
```

Comment:
This method looks more like a computed property to me. A common pattern in our codebase is to have an underlying stored property that holds the value, and a computed property that controls logic like this. You'll want to rename the `SESSION_ID` property with `_SESSION_ID`, then replace this with:

```
    private var SESSION_ID: String {
        get {
            guard let sessionID = _SESSION_ID else {
                let newID = generate_id()
                _SESSION_ID = newID
                return newID
            }
            
            return sessionID
        }
    }
```

Then you can replace `session_id()` method calls with a cleaner `SESSION_ID` property. Note you will also have to change `SESSION_ID = nil` to `_SESSION_ID = nil` in `begin_new_session()`.

Reply:
Thank you, this is much better